File: python/telco_system/telco_system_tests.py
from nose.tools import istest, eq_
from telco_system import TelcoSystem
from testing_utils import *

@istest
def the_historical_total_billed_should_be_zero_if_no_calls_were_made():
    telco_system = TelcoSystem()

    eq_(0, telco_system.historical_total_billed())


@istest
def an_international_call_of_two_minutes_should_cost_4_pesos():
    telco_system = TelcoSystem()

    telco_system.register_international_call_between(a_date_time, increment(a_date_time, in_minutes=2), "Juan")

    eq_(4, telco_system.historical_total_billed())
    eq_(1, telco_system.historical_number_of_calls())


@istest
def a_national_call_of_100_minutes_should_cost_150_pesos():
    telco_system = TelcoSystem()
    telco_system.register_national_call_between(a_date_time, increment(a_date_time, in_minutes=100), "Juan")

    eq_(150, telco_system.historical_total_billed())
    eq_(1, telco_system.historical_number_of_calls())


@istest
def a_local_call_of_10_minutes_during_peak_hour_should_cost_10_pesos():
    telco_system = TelcoSystem.with_peak_hour_starting_at(18)
    eighteen_o_clock = a_date_time_at(hour=18, minutes=0)

    telco_system.register_local_call_between(eighteen_o_clock, increment(eighteen_o_clock, in_minutes=10), "Juan")

    eq_(10, telco_system.historical_total_billed())
    eq_(1, telco_system.historical_number_of_calls())


@istest
def a_local_call_of_6_minutes_during_not_peak_hour_should_cost_3_pesos():
    telco_system = TelcoSystem.with_peak_hour_starting_at(18)
    two_o_clock = a_date_time_at(hour=14, minutes=0)

    telco_system.register_local_call_between(two_o_clock, increment(two_o_clock, in_minutes=6), "Juan")

    eq_(3, telco_system.historical_total_billed())
    eq_(1, telco_system.historical_number_of_calls())


@istest
def a_local_call_started_in_peak_hour_and_finished_in_non_peak_hour_should_consider_each_cost():
    telco_system = TelcoSystem.with_peak_hour_starting_at(18)
    two_minutes_before_peak_hour = a_date_time_at(hour=17, minutes=58)

    telco_system.register_local_call_between(two_minutes_before_peak_hour, increment(two_minutes_before_peak_hour, in_minutes=6), "Juan")

    eq_(5, telco_system.historical_total_billed())
    eq_(1, telco_system.historical_number_of_calls())


@istest
def two_international_calls_of_2_minutes_should_cost_8_pesos_for_juan():
    telco_system = TelcoSystem()

    telco_system.register_international_call_between(a_date_time, increment(a_date_time, in_minutes=2), "Juan")
    telco_system.register_international_call_between(another_date_time, increment(another_date_time, in_minutes=2), "Juan")

    telco_system.register_international_call_between(a_date_time, increment(a_date_time, in_minutes=20), "Pedro")
    telco_system.register_international_call_between(another_date_time, increment(another_date_time, in_minutes=12), "Pedro")

    eq_(8, telco_system.total_billed_for("Juan"))

    eq_(2, telco_system.number_of_calls_for("Juan"))
    eq_(4, telco_system.historical_number_of_calls())

File: c#/telco_system/TelcoSystem/TelcoSystemTest/TelcoSystemTests.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using TelcoSystem;

namespace TelcoSystemTest
{
    [TestClass]
    public class TelcoSystemTests
    {
        private DateTime aDateTime = DateTime.Now;
        private DateTime anotherDateTime = DateTime.Now.AddDays(2);
        private double delta = 0.01;

        [TestMethod]
        public void TheHistoricalTotalBilledShouldBeZeroIfNoCallsWereMade() {
            Telco telcoSystem = new Telco();

            Assert.AreEqual(0, telcoSystem.HistoricalTotalBilled(), delta);
        }

        [TestMethod]
        public void AnInternationalCallOfTwoMinutesShouldCost4Pesos() {
            Telco telcoSystem = new Telco();

            telcoSystem.RegisterInternationalCallBetween(aDateTime, aDateTime.AddMinutes(2), "Juan");

            Assert.AreEqual(4, telcoSystem.HistoricalTotalBilled(), delta);
            Assert.AreEqual(1, telcoSystem.HistoricalNumberOfCalls());
        }

        [TestMethod]
        public void TestANationalCallOf100MinutesShouldCost150Pesos() {
            Telco telcoSystem = new Telco();
            telcoSystem.RegisterNationalCallBetween(aDateTime, aDateTime.AddMinutes(100), "Juan");

            Assert.AreEqual(150, telcoSystem.HistoricalTotalBilled(), delta);
            Assert.AreEqual(1, telcoSystem.HistoricalNumberOfCalls());
        }

        [TestMethod]
        public void TestALocalCallOf10MinutesDuringRushHourShouldCost10Pesos() {
            Telco telcoSystem = Telco.WithRushHourStartingAt(18);
            DateTime eighteenOClock = new DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day, 18, 0, 0);

            telcoSystem.RegisterLocalCallBetween(eighteenOClock, eighteenOClock.AddMinutes(10), "Juan");

            Assert.AreEqual(10, telcoSystem.HistoricalTotalBilled(), delta);
            Assert.AreEqual(1, telcoSystem.HistoricalNumberOfCalls());
        }

        [TestMethod]
        public void TestALocalCallOf6MinutesDuringNotRushHourShouldCost3Pesos() {
            Telco telcoSystem = Telco.WithRushHourStartingAt(18);
            DateTime twoOClock = new DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day, 14, 0, 0);

            telcoSystem.RegisterLocalCallBetween(twoOClock, twoOClock.AddMinutes(6), "Juan");

            Assert.AreEqual(3, telcoSystem.HistoricalTotalBilled(), delta);
            Assert.AreEqual(1, telcoSystem.HistoricalNumberOfCalls());
        }

        [TestMethod]
        public void TestALocalCallStartedInRushHourAndFinishedInNonRushHourShouldConsiderEachCost() {
            Telco telcoSystem = Telco.WithRushHourStartingAt(18);
            DateTime twoMinutesBeforeRushHour = new DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day, 17, 58, 0);

            telcoSystem.RegisterLocalCallBetween(twoMinutesBeforeRushHour, twoMinutesBeforeRushHour.AddMinutes(6), "Juan");

            Assert.AreEqual(5, telcoSystem.HistoricalTotalBilled(), delta);
            Assert.AreEqual(1, telcoSystem.HistoricalNumberOfCalls());
        }

        [TestMethod]
        public void TestTwoInternationalCallsOf2MinutesShouldCost8PesosForJuan() {
           Telco telcoSystem = new Telco();

            telcoSystem.RegisterInternationalCallBetween(aDateTime, aDateTime.AddMinutes(2), "Juan");
            telcoSystem.RegisterInternationalCallBetween(anotherDateTime, anotherDateTime.AddMinutes(2), "Juan");

            telcoSystem.RegisterInternationalCallBetween(aDateTime, aDateTime.AddMinutes(20), "Pedro");
            telcoSystem.RegisterInternationalCallBetween(anotherDateTime, anotherDateTime.AddMinutes(12), "Pedro");

            Assert.AreEqual(8, telcoSystem.TotalBilledFor("Juan"), delta);

            Assert.AreEqual(2, telcoSystem.NumberOfCallsFor("Juan"));
            Assert.AreEqual(4, telcoSystem.HistoricalNumberOfCalls());
        }
    }
}
